wizard.py

```diff
@@ -0,0 +1,1038 @@
+  for auth_item in auth_items:
+    auth_dict[auth_item["step_name"]] = auth_item
+  return auth_dict
+
+
+# -----------------------------------------------------------------------------
+# NoSQL Helpers
+# -----------------------------------------------------------------------------
+# NOTE: Below functions rely on the current session having an id. If there is
+# no id in the session, all functions redirect to `404` (page not found).
+# This should never happen because all calling views should be decorated with
+# @with_session_id, which ensures that the current session does have an id.
+def _persist_session_subdocument(subdocument):
+  """Update a subdocument (e.g. vcs, ssc, functionaries...) in session document
+  identified by current session id. """
+  if not session.get("id"):
+    abort(404)
+
+  # Search session document by session ID in DB and update (replace)
+  # subdocument. If the entire document does not exist it is inserted
+  mongo.db.session_collection.update_one(
+    {"_id": session["id"]},
+    {"$set": subdocument},
+    upsert=True)
+
+
+def _persist_session_subdocument_ts(subdocument):
+  """Updates/adds last_modified to the subdocument before persisting it. """
+
+  for key in subdocument.keys():
+    subdocument[key]["last_modified"] = time.time()
+  _persist_session_subdocument(subdocument)
+
+
+def _get_session_subdocument(key):
+  """Returns a subdocument (e.g. vcs, ssc, functionaries...) identified by
+  passed key from session document identified by current session id.
+  Returns an empty dict if document or subdocument are not found.  """
+  if not session.get("id"):
+    abort(404)
+
+    # Get session document (use short circuit for default empty dict)
+  session_doc = mongo.db.session_collection.find_one(
+      {"_id": session["id"]})
+
+  if not session_doc:
+    return {}
+
+  # Get vcs data from session document or empty dict
+  return session_doc.get(key, {})
+
+
+def _get_session_document():
+  """Returns the entire session document.
+  Returns an empty dict if document or subdocument are not found.  """
+
+  if not session.get("id"):
+    abort(404)
+
+  session_doc = mongo.db.session_collection.find_one(
+    {"_id": session["id"]})
+
+  if not session_doc:
+    return {}
+
+  return session_doc
+
+
+# -----------------------------------------------------------------------------
+# View Decorator & Hooks
+# -----------------------------------------------------------------------------
+def with_session_id(wrapped_func):
+  """
+  Generate new session id if it does not exist.
+
+  For now, a user could start a new session on any page
+  TODO: Should we redirect to the start page if the session is new?
+  """
+  @wraps(wrapped_func)
+  def decorated_function(*args, **kwargs):
+    if not session.get("id"):
+      # Sessions use a MD5 hexdigest of a random value
+      # Security is not paramount, we don't store sensitive data, right?
+      session["id"] = hashlib.md5(str(random.random())).hexdigest()
```

Comment:
I wonder if you'd rather use pythons uuid module.

Reply:
Also, flask doesn't have session management? What does this framework even do then? 👎

Reply:
No need to get emotional! :P There is a Flask-Session extension, which i didn't find necessary, it uses [`uuid.uuid4`](https://pythonhosted.org/Flask-Session/#flask.ext.session.ServerSideSession.sid) for session ids. Will adopt!

Reply:
thanks!